// @fitbit/sdk/lib/buildTargets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateBuildTargets = void 0;
const tslib_1 = require("tslib");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const semver_1 = tslib_1.__importDefault(require("semver"));
const sdkVersion_1 = tslib_1.__importDefault(require("./sdkVersion"));
const baseBuildTargets = {
    atlas: {
        displayName: 'Fitbit Versa 3',
        minSDKVersion: '5.0.0',
        platform: ['128.1.1+'],
        resourceFilterTag: '336x336',
        specs: {
            screenSize: {
                width: 336,
                height: 336,
            },
        },
    },
    vulcan: {
        displayName: 'Fitbit Sense',
        minSDKVersion: '5.0.0',
        platform: ['128.1.1+'],
        resourceFilterTag: '336x336',
        specs: {
            screenSize: {
                width: 336,
                height: 336,
            },
        },
    },
};
let extraBuildTargets;
try {
    extraBuildTargets = require('@fitbit/sdk-build-targets').default;
}
catch (_a) { }
function generateBuildTargets() {
    return lodash_1.default.pickBy(Object.assign(Object.assign({}, baseBuildTargets), extraBuildTargets), ({ minSDKVersion }) => minSDKVersion === undefined ||
        semver_1.default.gte(sdkVersion_1.default().format(), minSDKVersion));
}
exports.generateBuildTargets = generateBuildTargets;
const buildTargets = generateBuildTargets();
exports.default = buildTargets;
